include/osquery/expected.h

```diff
@@ -10,131 +10,190 @@
  * auto result = function();
  * if (result) {
  *   ...use *result
  * } else {
- *   auto error = result->getError();
+ *   switch (result.getErrorCode()) {
+ *     case TestError::SomeError:
+ *        ...do something with it
+ *        break;
+ *     case TestError::AnotherError:
+ *        ...do something with it
+ *        break;
+ *   }
  * }
+ * @see osquery/core/tests/exptected_tests.cpp for more examples
  */
 
 namespace osquery {
 
-template <class ErrorCodeEnumType>
+template <typename ValueType_, typename ErrorCodeEnumType>
 class Expected final {
  public:
-  Expected(ErrorCodeEnumType object)
-      : object_(std::move(object)), hasError_(false) {}
-  Expected(ErrorBase* error) = delete;
-  Expected() : error_(nullptr), hasError_(false) {}
-  Expected(std::shared_ptr<ErrorBase> error)
-      : error_(std::move(error)), hasError_(true) {}
-  Expected(std::unique_ptr<ErrorBase> error)
-      : error_(std::move(error)), hasError_(true) {}
-  template <class ErrorT>
-  Expected(std::shared_ptr<Error<ErrorT>> error)
-      : error_(std::static_pointer_cast<ErrorBase>(error)), hasError_(true){};
+  using ValueType = ValueType_;
+  using ErrorType = Error<ErrorCodeEnumType>;
+  using SelfType = Expected<ValueType, ErrorCodeEnumType>;
 
+ public:
+  Expected(ValueType value) : object_{std::move(value)} {}
+
+  Expected(ErrorType error) : object_{std::move(error)} {}
+
+  explicit Expected(ErrorCodeEnumType code, std::string message)
+      : object_{ErrorType(code, message)} {}
+
+  Expected(Expected&& other) = default;
+
+  Expected() = delete;
   Expected(const Expected&) = delete;
+  Expected(ErrorBase* error) = delete;
+
+  Expected& operator=(Expected&& other) = default;
+  Expected& operator=(const Expected& other) = delete;
 
   ~Expected() {
     assert(errorChecked_ && "Error was not checked");
   }
 
-  Expected& operator=(Expected&& other) {
-    if (this != &other) {
-      object_ = std::move(other.object_);
-      error_ = std::move(other.error_);
-      hasError_ = other.hasError_;
-    }
-    return *this;
+  static SelfType success(ValueType value) {
+    return SelfType{std::move(value)};
   }
 
-  Expected(Expected&& other) {
-    object_ = std::move(other.object_);
-    error_ = std::move(other.error_);
-    hasError_ = other.hasError_;
+  static SelfType failure(std::string message) {
+    auto defaultCode = ErrorCodeEnumType{};
+    return SelfType(defaultCode, std::move(message));
   }
 
-  std::shared_ptr<ErrorBase> getError() const {
-    return error_;
+  static SelfType failure(ErrorCodeEnumType code, std::string message) {
+    return SelfType(code, std::move(message));
   }
 
-  explicit operator bool() const {
+  ErrorType takeError() {
+    return std::move(boost::get<ErrorType>(object_));
+  }
+
+  const ErrorType& getError() const {
+    return boost::get<ErrorType>(object_);
+  }
+
+  ErrorCodeEnumType getErrorCode() const {
+    return getError().getErrorCode();
+  }
+
+  bool isOk() const noexcept {
```

Comment:
can you replace it with something more explisit like hasError() or remove it all to avoid providing 2 methods to do the same. IMHO good example is boost::optional that does not have other methods to check values expect bool operator()

Reply:
Yep, definitely it makes sense, thanks!